bpf/generictracer/protocol_kafka.h

```diff
@@ -0,0 +1,523 @@
+// Given the fact that we need to store this for the duration of the full request
+// (split in potentially multiple packets), we will **not** process or preserve
+// any actual payloads that are exactly 4 bytes long — they are intentionally
+// dropped in favor of state storage.
+static __always_inline int kafka_store_state_data(const connection_info_t *conn_info,
+                                                  const unsigned char *data,
+                                                  size_t data_len,
+                                                  u8 direction) {
+
+    // we want to store only request/response of split sends that are 4 bytes long
+    if (data_len != k_kafka_hdr_message_size) {
+        return 0;
+    }
+
+    int message_size = kafka_read_message_size(data, data_len);
+    if (message_size == -1) {
+        return 0;
+    }
+    struct kafka_state_data new_state_data = {};
+    new_state_data.message_size = message_size;
+    struct kafka_state_key state_key = {.conn = *conn_info, .direction = direction};
+    bpf_map_update_elem(&kafka_state, &state_key, &new_state_data, BPF_ANY);
+
+    return -1;
+}
+
+// Request header
+// +--------------+-----------------+---------------------+----------------|
+// | message_size | request_api_key | request_api_version | correlation_id |
+// +--------------+-----------------+---------------------+----------------|
+// |    4B        |     2B          |     2B              |      4B        |
+// +--------------+-----------------+---------------------+----------------|
+static __always_inline int kafka_parse_fixup_request_header(const connection_info_t *conn_info,
+                                                            struct kafka_request_hdr *hdr,
+                                                            const unsigned char *data,
+                                                            size_t data_len,
+                                                            u8 direction) {
+
+    // Try to parse and validate the header first.
+    bpf_probe_read(&hdr->message_size, k_kafka_hdr_message_size, (const void *)data);
+    hdr->message_size = bpf_ntohl(hdr->message_size);
+    if (hdr->message_size == (data_len - k_kafka_hdr_message_size)) {
+        bpf_probe_read(&hdr->request_api_key,
+                       k_kafka_hdr_request_api_key,
+                       (const void *)(data + k_kafka_hdr_message_size));
+        hdr->request_api_key = bpf_ntohs(hdr->request_api_key);
+        if (!is_kafka_api_key_supported(hdr->request_api_key)) {
+            bpf_dbg_printk("kafka_parse_fixup_request_header: request_api_key provided %d, is "
+                           "not supported",
+                           hdr->request_api_key);
+            return -1;
+        }
+        bpf_probe_read(
+            &hdr->request_api_version,
+            k_kafka_hdr_request_api_version,
+            (const void *)(data + k_kafka_hdr_message_size + k_kafka_hdr_request_api_key));
```

Comment:
should we keep an advancing offset instead of adding constants inline?

Reply:
you mean like constants with the incremented value? Like 
`k_kafka_hdr_message_size = 4`, `k_kafka_hdr_message_size_and_key = 6`, `k_kafka_hdr_message_size_key_and_version= 8`...?